### DISCUSSION TITLE
Write Queue size

### DISCUSSION THREAD

**Discussion by rozza591** [bluenviron/mediamtx#2940] [Questions]
### Question

Hey, what would people say an appropriate write queue size was. Is the size directory related to ram usage. IE: If I set the write queue size to 8,192 would that use up to 8gb of ram. 

I'm asking as I keep reveiving a write queue full warning but I've assigned it 16,384 thinking that would allow it to us 16gb of ram. I'm no where close to using that much ram and still receive the error. I've got 64gb in total so do I just allow it to use something stupid like 50gb or is that just stupid.

**Comment by cubisttriangle** [comment]
What are you sending to mediamtx? It seems like your input stream might have something wrong or unusual.

**Reply by rozza591** [reply]
It was an rtsp stream pushed by ffmpeg (copy, no transcoding). The source was a concatonated list of 3mbs h264 videos with no audio.

**Comment by aler9** [comment]
Hello, in my opinion you have to touch `writeQueueSize` only when warnings about packet corruption start to appear on client side, or "RTP packet lost" warnings start to appear on server side. In this case, increase it until the messages disappear or are greatly reduced in number.

**Reply by Arcitec** [reply]
The warnings in the server log look like this: `WAR [RTSP] [session 11a4edfc] write queue is full`. And on the player/client I see errors about corrupt videos (since chunks are being lost).

And these errors are easily fixed by modifying `writeQueueSize` in the MediaMTX configuration.

I transmit a 15 Mbit stream and got rid of the errors by starting the container like this (because all config file settings can be overridden via environment variables instead of editing `mediamtx.yml`):

`docker run --rm -it --network=host -e MTX_WRITEQUEUESIZE=2048 docker.io/bluenviron/mediamtx:latest`

What I've understood so far is the "write queue" is the circular buffer that stores the incoming video chunks. It [was described in this old 2022 thread](https://github.com/bluenviron/mediamtx/issues/1046#issuecomment-1206603549) (back then, the same setting used to be named `readBufferCount` instead, but has been renamed now). Also note that [the advice about switching to TCP is outdated](https://github.com/bluenviron/mediamtx/discussions/356#discussioncomment-621594) and was fixed in 2021 so that MediaMTX handles both TCP and UDP very well, but TCP *can* still be more reliable for ingress due to the OS buffering packets for you if the write queue is full. But UDP (what I use) also works if you just raise the write queue to be big enough!

The sender sends into that circular buffer, and the clients all read from the buffer.

By increasing the size of the buffer, it can fit more data.

It needs to be large enough to fit all the data of the incoming stream's bitrate.

The default value is `512` which is safe for streams up to 5Mbit, or maybe a bit higher (maybe up to 6.5 Mbit?). I saw the "write queue is full" error occasionally appear when I transmitted at 8Mbit (the actual rate reported in FFmpeg was around 7.4 Mbit). And it appeared all the time when I transmitted at 15Mbit.

Raising the `writeQueueSize` to `1024` allowed me to do 15Mbit H264 streaming without the errors, but I don't trust it long-term since `512` struggles around 7.4-8Mbit, so `1024` is definitely "on the edge" of enough and might fail sometimes.

**I'll set it to `MTX_WRITEQUEUESIZE=2048` now just to be totally safe. It would also be very safe to go a bit lower, but the MediaMTX option itself only supports "power of 2" values such as 512, 1024, 2048, etc, so I am forced to pick this value.**

**Update: I have spotted one more situation that can cause the `write queue is full` message:** If a viewer pauses their video player, so that they aren't requesting more data from the server anymore, then you will see `write queue is full` every second in the server logs until they unpause the stream. This can be safely ignored. It's only a worry if the write queue is full when the vide encoder transmits data to MediaMTX server. You don't need to care about the warnings when the users pause the video. However, I suspect that it can also happen if a viewer is connected via TCP and has a slow connection, which means the server can't write the data fast enough to that client. In that case, it can also be ignored, or you can lower the video's bitrate to accommodate more people. :)

**Reply by Arcitec** [reply]
@rozza591 Your numbers were definitely too high by the way, so I suspect that your "3Mbit files" are corrupt or getting broken in a different way (either the input is broken or your MediaMTX ingress protocol is unstable).

I transmit 15Mbit H264 via SRT (UDP) protocol, and writeQueueSize 1024 was okay but 2048 is what I use to be 100% safe.

**Comment by yelodevopsi** [comment]
Just to flip the question on it's head:
* What would it hurt if you set a too high value? (i.e. we're debugging some stuff setting it to 8192)
* Would it only be more RAM-usage?
* Other (bad) side-effects?

**Reply by pikachu937** [reply]
Hello everyone and as always specially to the great and mighty **aler9**.
I would also like to get a clear answer to these questions, as well as to the main question "how to correctly calculate the required value".
I have N servers on which 1000 streams are dynamically published per server, all streams are published only via rtsp with udp transport, there are servers on which only 720p streams, there are servers with 1080p streams, and there are servers with 1440p and 2160p streams, and there are also hybrid servers on which both 720p and 1080p and 1440p and 2160p streams are collected. I have a lot of messages like "RTP packet lost", and on some servers I come across messages like "write queue is full".

I will immediately make a reservation, each server has 128G RAM, my network is 10G and there are no problems in the network, I am 146% sure.

**Please help me and answer the question how can I correctly calculate what value needs to be set on a particular server?**

Thank you in advance.
